src/lib.rs

```diff
@@ -0,0 +1,218 @@
+extern crate regex;
+use regex::Regex;
+use std::error;
+use std::fmt;
+
+pub struct Siso {
+    route: Route,
+}
+
+impl Siso {
+    pub fn register(
+        &mut self,
+        path_nodes: &[PathNode],
+        value: RouteValue,
+    ) -> Result<(), SisoError> {
+        if path_nodes.is_empty() {
+            panic!("Need at least one node for registering a value...");
+        }
+        self.route.register(path_nodes, value)
+    }
+
+    pub fn find(&self, path_nodes: &[PathNode]) -> Result<RouteValue, SisoError> {
+        if path_nodes.is_empty() {
+            panic!("Need at least one node for finding a value...");
+        }
+        self.route.find(path_nodes)
+    }
+
+    pub fn new() -> Siso {
+        Siso {
+            route: Route::new(PathNode::PathPart(String::from("[root]")), None),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Route {
+    node: PathNode,
+    routes: Vec<Route>,
+    value: RouteValue,
+}
+
+impl Route {
+    fn register(&mut self, path_nodes: &[PathNode], value: RouteValue) -> Result<(), SisoError> {
+        let path_node = &path_nodes[0];
+        let is_leaf_path_node = path_nodes.len() == 1;
+        let index = self.routes.iter()
```

Comment:
@matklad Find out a way to get the index more nicely but still stucked on relying on the index for muting it later. Will stay with it and go further in my experiments. Thanks for your review!